### PR TITLE
First in a series on API development I'd like to do

### PR DESCRIPTION
**Pull request by dferguson992**

### CODE REVIEW

File: posts/using-swagger-to-build-effective-apis.md [hunk 1]
```diff
@@ -0,0 +1,382 @@
...
+
+Pathing is a complex topic in API development.  There are many things to consider when defining API paths, like HTTP Headers, HTTP Methods, endpoint names, response types, versions, etc.  For example, you will notice each API path I have specified is a noun.  This noun is either singular or plural, depending on the number of objects returned by the response definition.  The only verbs in this API design are the HTTP verbs, which is as it should be.  Some developers may be tempted to say `http://www.myapi.com/getAllExercises`.  This is not good practice; behind the scenes this operations is a `GET getAllExercises`, which is redundant.  Instead, define your API paths as nouns and use the HTTP operations as verbs.  If you can concatenate the two and create a sensible phrase, you have successfully defined a logical endpoint like `POST exercise` or `GET exercises`.
```
Comment (aseigneurin) [inline posts/using-swagger-to-build-effective-apis.md]: Agree. You could say that a `getAllExercises` endpoint would be RPC, not REST.

File: posts/using-swagger-to-build-effective-apis.md [hunk 2]
```diff
@@ -0,0 +1,382 @@
...
+      "repetitionModified" : "tempo"
+    } ],
+    "name" : "string"
+  } ],
+  "avgRestTime" : 0
+}
+```
+As you can see, these two GET requests return basic Exercise objects.  This mocking allows us to view the exercises as they would be returned from the API.  This is important, as it allows us to easily define improved endpoints for future releases of the API.  Seeing the results of the endpoint requests, even if they are mocked, will always be more valuable than writing the specification and never seeing it in action until business logic is written.
+
+We can even modify the request body of the POST endpoint we defined earlier
```
Comment (aseigneurin) [inline posts/using-swagger-to-build-effective-apis.md]:
```suggestion
We can even modify the request body of the POST endpoint we defined earlier:
```

File: posts/using-swagger-to-build-effective-apis.md [hunk 3]
```diff
@@ -0,0 +1,382 @@
...
...ould always adhere to a common standard which will make your API usable.  That is why you should use the OAS 3.0.n specification for your API.  APIs that do not adhere to this design have the potential to be difficult to use and maintain.  Any tool that is difficult to maintain will not be used.  Inversely, Swagger is a tool that is very easy to use, which is why I have developed this demo API using Swagger tools in the first place.  Some of the additional benefits of Swagger tools include their portability.  Swagger Hub will let you generate templated client and server code in nearly any language you want.  This allows you to rapidly prototype APIs for the future.  AWS API Gateway will generate help you to test your API, or even deploy your API in a serverless context using AWS Lambda to handle endpoint calls.  Maven has a swagger plug-in which takes a definition YAML file and generates client models on the fly so you can rapidly prototype an application for your API.  If the API changes, you just update the definition file and your API code will change with it.  Swagger is the toolset for helping architects write APIs quickly, and OAS 3.0 allows these APIs to be used be everyone.
```
Comment (aseigneurin) [inline posts/using-swagger-to-build-effective-apis.md]: `AWS API Gateway will generate help you to test your API ` ?

File: posts/using-swagger-to-build-effective-apis.md [hunk 4]
```diff
@@ -0,0 +1,382 @@
...
+- Dan Ferguson
+tags:
+- API
+- Swagger
+date: 2019-02-07T20:22:38.000Z
+title: "Using Swagger to Build Effective APIs"
+image:
+---
+
+The rise of serverless computing and platform agnostic, app-based service use has forced the casual developer to learn proper API development, a task usually reserved for platform architects.  Unfortunately, there are many programmers, start-ups, and full-fledged companies that do not adhere to proper API development and maintenance.  The skillsets and mindsets involved in creating APIs are different from those involved in developing the business logic which uses an API.  This schism does not affect developers or architects; it only affects the users of the API.  Poorly developed or poorly documented APIs which are susceptible frequent changes or misleading design decisions are difficult to use.  If a product is difficult to use it likely will not be used at all, especially if that product is an API.  The good news is there are specifications and tools which aide in the development of easy to use APIs that change as fast as your business.  I am speaking specifically about the Open API Specification and Swagger.
```
Comment (aseigneurin) [inline posts/using-swagger-to-build-effective-apis.md]:
```suggestion
The rise of serverless computing and platform agnostic, app-based services use has forced the casual developer to learn proper API development, a task usually reserved for platform architects.  Unfortunately, there are many programmers, start-ups, and full-fledged companies that do not adhere to proper API development and maintenance.  The skillsets and mindsets involved in creating APIs are different from those involved in developing the business logic which uses an API.  This schism does not affect developers or architects; it only affects the users of the API.  Poorly developed or poorly documented APIs which are susceptible frequent changes or misleading design decisions are difficult to use.  If a product is difficult to use it likely will not be used at all, especially if that product is an API.  The good news is there are specifications and tools which aide in the development of easy to use APIs that change as fast as your business.  I am speaking specifically about the Open API Specification and Swagger.
```

File: posts/using-swagger-to-build-effective-apis.md (hunk 3 above)
Comment (dferguson992) [reply posts/using-swagger-to-build-effective-apis.md]: Ah yeah that's an awkward statement.  I meant to say that if you decide to use AWS API Gateway to manage/deploy your API, it's much easier if you already have a Swagger document.  

Will adjust.

File: posts/using-swagger-to-build-effective-apis.md (hunk 1 above)
Comment (dferguson992) [reply posts/using-swagger-to-build-effective-apis.md]: Good point, modified this piece to add a nice article which breaks down the difference between REST and RPC

File: posts/using-swagger-to-build-effective-apis.md [hunk 5]
```diff
@@ -0,0 +1,390 @@
...
+
+Take a look at this graphic:
+
+![](https://github.com/ippontech/blog-usa/blob/master/images/2019/02/aspotr_umls.png)
+
+Here we define a workout as an array of exercises.  Each exercise is a combination of sets with a rest time.  Each set is an array of repetitions defined primarily by a weight, as well as a lift, which is defined primarily by an affected muscle group.  Now we have the framework to easily define a weight lifting workout.  How do we translate this diagram of loose fitting parts into an API?  There are three steps we will need to take to turn these UML diagrams into tangible JSON packets.  The first step is to set up our local Swagger development environment.  The second is to populate our API definitions with some sample paths.  The last step is to test these sample paths using Swagger's testing tools on Swagger Hub.  It is important to note that because we are using Swagger tools to develop our API, we will automatically be prescribing to OAS 3.0 standards, ensuring our API will be easy to use for developers.
```
Comment (yhippa) [inline posts/using-swagger-to-build-effective-apis.md]:
```suggestion
Here we define a workout as an array of exercises.  Each exercise is a combination of sets with a rest time.  Each set is an array of repetitions defined primarily by a weight, as well as a lift, which is defined primarily by an affected muscle group.  Now we have the framework to easily define a weight lifting workout.  How do we translate this diagram of loose fitting parts into an API?

There are three steps we will need to take to turn these UML diagrams into tangible JSON packets.  The first step is to set up our local Swagger development environment.  The second is to populate our API definitions with some sample paths.  The last step is to test these sample paths using Swagger's testing tools on Swagger Hub.  It is important to note that because we are using Swagger tools to develop our API, we will automatically be prescribing to OAS 3.0 standards, ensuring our API will be easy to use for developers.
```

File: posts/using-swagger-to-build-effective-apis.md [hunk 6]
```diff
@@ -0,0 +1,390 @@
...
+```YAML
+# Added by API Auto Mocking Plugin
+  - description: SwaggerHub API Auto Mocking
+    url: https://virtserver.swaggerhub.com/dferguson992/aspotr/1.0.0
+```
+This is a virtual server used to mock requests to your API.  Your username, API name, and API Version number all define the URL of the mock server.  You cannot ping the server, but you can run curl commands against API endpoints hosted on the server.  These API endpoints are pulled straight from your specification,  and will send mock data to your endpoints so you can physically see the responses.  Furthermore, you are free to modify the mock data to fit any edge cases you may want to program into your application in the future.  It is important to note, this is just a sanity test designed to allow you to see the specified output of your API.  You should not use the mock endpoints to actually test your business logic.  The mock endpoint is created solely for the purpose of viewing and verifying expected output.
```
Comment (yhippa) [inline posts/using-swagger-to-build-effective-apis.md]:
```suggestion
This is a virtual server used to mock requests to your API.  Your username, API name, and API Version number all define the URL of the mock server.  You cannot ping the server, but you can run curl commands against API endpoints hosted on the server.  These API endpoints are pulled straight from your specification and will send mock data to your endpoints so you can physically see the responses.  Furthermore, you are free to modify the mock data to fit any edge cases you may want to program into your application in the future.  It is important to note, this is just a sanity test designed to allow you to see the specified output of your API.  You should not use the mock endpoints to actually test your business logic.  The mock endpoint is created solely for the purpose of viewing and verifying the expected output.  
```

File: posts/using-swagger-to-build-effective-apis.md [hunk 7]
```diff
@@ -0,0 +1,390 @@
...
...re to this design have the potential to be difficult to use and maintain.  Any tool that is difficult to maintain will not be used.  Inversely, Swagger is a tool that is very easy to use, which is why I have developed this demo API using Swagger tools in the first place.  Some of the additional benefits of Swagger tools include their portability.  Swagger Hub will let you generate templated client and server code in nearly any language you want.  This allows you to rapidly prototype APIs for the future.  For example, AWS API Gateway will generate all of the features of your OAS compliant API like models and routes, will help you test your API, and will even deploy your API in a serverless context using AWS Lambda to handle endpoint calls.  The only thing API Gateway needs to get started is a Swagger document.  Maven has a swagger plug-in which takes a definition YAML file and generates client models on the fly so you can rapidly prototype an application for your API.  If the API changes, you just update the definition file and your API code will change with it.  Swagger is the toolset for helping architects write APIs quickly, and OAS 3.0.n allows these APIs to be used be everyone.
```
Comment (yhippa) [inline posts/using-swagger-to-build-effective-apis.md]:
```suggestion
Swagger is a great toolset for anyone looking to develop an API.  For those of us looking to write APIs, you should always adhere to a common standard which will make your API usable.  That is why you should use the OAS 3.0.n specification for your API.  APIs that do not adhere to this design have the potential to be difficult to use and maintain.  Any tool that is difficult to maintain will not be used.

Inversely, Swagger is a tool that is very easy to use, which is why I have developed this demo API using Swagger tools in the first place.  Some of the additional benefits of Swagger tools include their portability.  Swagger Hub will let you generate templated client and server code in nearly any language you want.  This allows you to rapidly prototype APIs for the future.

For example, AWS API Gateway will generate all of the features of your OAS compliant API like models and routes, will help you test your API, and will even deploy your API in a serverless context using AWS Lambda to handle endpoint calls.  The only thing API Gateway needs to get started is a Swagger document.  Maven has a swagger plug-in which takes a definition YAML file and generates client models on the fly so you can rapidly prototype an application for your API.  If the API changes, you just update the definition file and your API code will change with it.  Swagger is the toolset for helping architects write APIs quickly, and OAS 3.0.n allows these APIs to be used by everyone.
```

File: posts/using-swagger-to-build-effective-apis.md [hunk 8]
```diff
@@ -0,0 +1,390 @@
...
+
+# Test your Swagger using Swagger Hub tools
+Once you've built your API Specification using the Swagger Editor, you have to test it.  By far, the easiest way you can test your API specification is to copy it from the Swagger Editor and paste it into Swagger Hub.  This will of course require you to sign-in to Swagger Hub and create a new API; but the ease of testing involved make the process well worth it.  Additionally, you can make any new API private on Swagger Hub, if you are concerned others may poach your API.
+
+Once you've pasted your API into the Swagger Hub, you'll notice the UI will take on a very similar look and feel as the Swagger Editor.  They are not identical displays however; in fact you may notice that Swagger Hub adds a server line for you.  This is what Swagger Hub added to the server section of my API
```
Comment (yhippa) [inline posts/using-swagger-to-build-effective-apis.md]:
```suggestion
Once you've pasted your API into the Swagger Hub, you'll notice the UI will take on a very similar look and feel as the Swagger Editor.  They are not identical displays however; in fact, you may notice that Swagger Hub adds a server line for you.  This is what Swagger Hub added to the server section of my API
```

File: posts/using-swagger-to-build-effective-apis.md [hunk 9]
```diff
@@ -0,0 +1,390 @@
...
+
+Usually, APIs are written by architects, for developers.  This immediately implies a paradigm shift will be required in order to keep your API future proof from developer requests.  If you want to design your API so it is easily upgradeable, easily maintanable, and easily understood by both users and developers, check out this cool articles on [API Design](https://hackernoon.com/restful-api-design-step-by-step-guide-2f2c9f9fcdbf) and [API Best Practices](https://hackernoon.com/restful-api-designing-guidelines-the-best-practices-60e1d954e7c9).
+
+# Test your Swagger using Swagger Hub tools
+Once you've built your API Specification using the Swagger Editor, you have to test it.  By far, the easiest way you can test your API specification is to copy it from the Swagger Editor and paste it into Swagger Hub.  This will of course require you to sign-in to Swagger Hub and create a new API; but the ease of testing involved make the process well worth it.  Additionally, you can make any new API private on Swagger Hub, if you are concerned others may poach your API.
```
Comment (yhippa) [inline posts/using-swagger-to-build-effective-apis.md]:
```suggestion
Once you've built your API Specification using the Swagger Editor, you have to test it.  By far, the easiest way you can test your API specification is to copy it from the Swagger Editor and paste it into Swagger Hub.  This will, of course, require you to sign-in to Swagger Hub and create a new API; but the ease of testing involved make the process well worth it.  Additionally, you can make any new API private on Swagger Hub, if you are concerned others may poach your API.
```

File: posts/using-swagger-to-build-effective-apis.md [hunk 10]
```diff
@@ -0,0 +1,390 @@
...
+
+Usually, APIs are written by architects, for developers.  This immediately implies a paradigm shift will be required in order to keep your API future proof from developer requests.  If you want to design your API so it is easily upgradeable, easily maintanable, and easily understood by both users and developers, check out this cool articles on [API Design](https://hackernoon.com/restful-api-design-step-by-step-guide-2f2c9f9fcdbf) and [API Best Practices](https://hackernoon.com/restful-api-designing-guidelines-the-best-practices-60e1d954e7c9).
```
Comment (yhippa) [inline posts/using-swagger-to-build-effective-apis.md]:
```suggestion
Usually, APIs are written by architects, for developers.  This immediately implies a paradigm shift will be required in order to keep your API future proof from developer requests.  If you want to design your API so it is easily upgradeable, easily maintainable, and easily understood by both users and developers, check out these articles: [API Design](https://hackernoon.com/restful-api-design-step-by-step-guide-2f2c9f9fcdbf) and [API Best Practices](https://hackernoon.com/restful-api-designing-guidelines-the-best-practices-60e1d954e7c9).
```

File: posts/using-swagger-to-build-effective-apis.md [hunk 11]
```diff
@@ -0,0 +1,390 @@
...
... versions, etc.  For example, you will notice each API path I have specified is a noun.  This noun is either singular or plural, depending on the number of objects returned by the response definition.  The only verbs in this API design are the HTTP verbs, which is as it should be.  Some developers may be tempted to say `http://www.myapi.com/getAllExercises`.  This is not good practice; behind the scenes this operations is a `GET getAllExercises`, which is redundant.  Instead, define your API paths as nouns and use the HTTP operations as verbs.  If you can concatenate the two and create a sensible phrase, you have successfully defined a logical endpoint like `POST exercise` or `GET exercises`.  The use of verbs to define REST API operations is a tricky business and is generally considered bad practice.  However, if you are defining a different kind of API, like SOAP or more generally RPC, this is good practice.  RPC APIs are beyond the scope of this post, but if you would like to do more reading, I would suggest [this article](https://www.smashingmagazine.com/2016/09/understanding-rest-and-rpc-for-http-apis/), which neatly defines the differences between REST APIs and RPC type APIs.
```
Comment (yhippa) [inline posts/using-swagger-to-build-effective-apis.md]:
```suggestion
Pathing is a complex topic in API development.  There are many things to consider when defining API paths, like HTTP Headers, HTTP Methods, endpoint names, response types, versions, etc.  For example, you will notice each API path I have specified is a noun.  This noun is either singular or plural, depending on the number of objects returned by the response definition.

The only verbs in this API design are the HTTP verbs, which is as it should be.  Some developers may be tempted to say `http://www.myapi.com/getAllExercises`.  This is not good practice; behind the scenes, this operation is a `GET getAllExercises`, which is redundant.  Instead, define your API paths as nouns and use the HTTP operations as verbs.

If you can concatenate the two and create a sensible phrase, you have successfully defined a logical endpoint like `POST exercise` or `GET exercises`.  The use of verbs to define REST API operations is a tricky business and is generally considered bad practice.  However, if you are defining a different kind of API, like SOAP or more generally RPC, this is good practice.  RPC APIs are beyond the scope of this post, but if you would like to do more reading, I would suggest [this article](https://www.smashingmagazine.com/2016/09/understanding-rest-and-rpc-for-http-apis/), which neatly defines the differences between REST APIs and RPC type APIs.
```

File: posts/using-swagger-to-build-effective-apis.md [hunk 12]
```diff
@@ -0,0 +1,390 @@
...
...choices.  The first is to slowly build your JSON workout object by periodically adding set objects to it throughout the day.  The second option is to send small workout requests objects throughout the day, each containing one exercise of about 20 or so push ups.  The first option is not convenient, but it lends itself well to exercise tracking, especially if you wanted to model the data you've collected and build progress projections over time.  In fact, this first approach is how one would assume any workout tracking is to be done.  The second approach is much more convenient for the user, even if the machine learning conducted on your data set is blown out of the water if multiple workouts of the same exercise are logged per day.  For our purposes now, this API will work just fine; but as you continue to read, consider optimizations to the API that would make it scalable and easy to use.  It is important to think about the API from all perspectives, those modeling the business logic on the back-end, and those submitted REST requests on the front-end.  If you keep these tenants in mind, you can build a much more atomic API that maintains its usability by both developers and users.
```
Comment (yhippa) [inline posts/using-swagger-to-build-effective-apis.md]:
```suggestion
It is important to note that the demo API in this post may not be a scalable API design.  For our purposes, this design may work just fine; it is glaringly obvious this design will create large, nested JSON payloads for even the shortest workouts.  The example API in this post implies each transaction is stateful, and payload components are dependent upon other potentially unrelated components.

For example, say you want to track the number of push-ups you've done in a day.  With this API design, you have two choices.  The first is to slowly build your JSON workout object by periodically adding set objects to it throughout the day.  The second option is to send small workout requests objects throughout the day, each containing one exercise of about 20 or so push-ups.

The first option is not convenient, but it lends itself well to exercise tracking, mainly if you wanted to model the data you've collected and build progress projections over time.  This first approach is how one would assume any workout tracking is to be done.  The second approach is much more convenient for the user, even if the machine learning conducted on your data set is blown out of the water if multiple workouts of the same exercise are logged per day.

For our purposes now, this API will work just fine; but as you continue to read, consider optimizations to the API that would make it scalable and easy to use.  It is essential to think about the API from all perspectives, those modeling the business logic on the back-end, and those submitted REST requests on the front-end.  If you keep these tenants in mind, you can build a much more atomic API that maintains its usability by both developers and users.
```

File: posts/using-swagger-to-build-effective-apis.md [hunk 13]
```diff
@@ -0,0 +1,390 @@
...
+          format: int64
+        exercises:
+          type: array
+          items:
+            $ref: '#/components/schemas/Exercise'
+      required:
+        - id
+        - exercises
+```
+Starting with our most primitive objects built on basic data types like MuscleGroup, we can define complex objects like Workout.  The glue that binds these objects together is the `$ref` tag.  This tag references another location in your API specification.  When defining models, we use it heavily to build our complex objects based on definitions of simpler objects.  This tag works by defining a structure similar to directories, where the `#` character is the root tag, and the `/` character seperates each level of the YAML document.  This structure helps to define components in your API using other, previously defined components.  This tag is incredibly powerful, as it allows you to not only reference components in this document, but to reference components in other API specification documents as well.  This is a great strategy is you are linking two or more extraneous APIs, or if you want to keep your component descriptions seperated from the rest of your API document.
```
Comment (yhippa) [inline posts/using-swagger-to-build-effective-apis.md]:
```suggestion
Starting with our most primitive objects built on basic data types like MuscleGroup, we can define complex objects like Workout.  The glue that binds these objects together is the `$ref` tag.  This tag references another location in your API specification.  When defining models, we use it heavily to build our complex objects based on definitions of simpler objects.  This tag works by defining a structure similar to directories, where the `#` character is the root tag, and the `/` character separates each level of the YAML document.

This structure helps to define components in your API using other, previously defined components.  This tag is incredibly powerful, as it allows you to not only reference components in this document and also reference components in other API specification documents as well.  This is a great strategy if you are linking two or more extraneous APIs, or if you want to keep your component descriptions separated from the rest of your API document.
```
